File: lib/py/uapi/internal/schema/FindSchemaKey.py
from collections import OrderedDict


def find_schema_key(document_name: str, definition: dict[str, object], index: int, document_names_to_json: dict[str, str]) -> str:
    from uapi.UApiSchemaParseError import UApiSchemaParseError
    from uapi.internal.schema.SchemaParseFailure import SchemaParseFailure
    import re

    regex = "^(((fn|errors|requestHeader|responseHeader|info)|((struct|union|_ext)(<[0-2]>)?))\\..*)"
    matches = []

    keys = sorted(list(definition.keys()))

    for e in keys:
        if re.match(regex, e):
            matches.append(e)

    if len(matches) == 1:
        return matches[0]
    else:
        parse_failure = SchemaParseFailure(document_name, [index],
                                           "ObjectKeyRegexMatchCountUnexpected",
                                           {"regex": regex, "actual": len(
                                               matches), "expected": 1, "keys": keys})
        raise UApiSchemaParseError([parse_failure], document_names_to_json)

File: lib/java/src/main/java/uapi/internal/schema/FindSchemaKey.java
package uapi.internal.schema;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import uapi.UApiSchemaParseError;

public class FindSchemaKey {

    public static String findSchemaKey(String documentName, Map<String, Object> definition, int index) {
        final var regex = "^(((fn|errors|requestHeader|responseHeader|info)|((struct|union|_ext)(<[0-2]>)?))\\..*)";
        final var matches = new ArrayList<String>();

        final var keys = definition.keySet().stream().sorted().toList();

        for (final var e : keys) {
            if (e.matches(regex)) {
                matches.add(e);
            }
        }

        if (matches.size() == 1) {
            return matches.get(0);
        } else {
            final var parseFailure = new SchemaParseFailure(documentName, List.of(index),
                    "ObjectKeyRegexMatchCountUnexpected",
                    new TreeMap<>(
                            Map.of("regex", regex, "actual", matches.size(), "expected", 1, "keys", keys)));
            throw new UApiSchemaParseError(List.of(parseFailure));
        }
    }
}
